File: Aio-Algorithms-And-Solutions/Solutions/Exercism/CSharp/Robot.cs
namespace Aio_Algorithms_And_Solutions.Solutions.Exercism.CSharp;

public class Robot
{
    private static readonly HashSet<string> Names = new ();
    private static readonly Random Random = new ();
    private string Name { get; set; }
    
    public Robot()
    {
        Reset();
    }

    public void Reset()
    {
        string newName;

        do
        {
            newName = GenerateName();
        }
        while (!Names.Add(newName));

        Names.Remove(Name);

        Name = newName;
    }

    private static string GenerateName()
    {
        var letterPart = (char)('A' + Random.Next(0, 26));
        var secondLetterPart = (char)('A' + Random.Next(0, 26));
        var numberPart = Random.Next(0, 1000); 
        return $"{letterPart}{secondLetterPart}{numberPart:D3}";
    }
}

File: Aio-Algorithms-And-Solutions/Solutions/Exercism/Java/Robot.java
package exercism;

import java.util.stream.Collectors;
import java.util.*;

public class Robot {
    private static final Set<String> registeredNames = new HashSet<String>();
    private static final Random random = new Random();
    private String name;
    public Robot() {
        reset();
    }
    public String getName() {
        return name;
    }
    public void reset() {
        registeredNames.remove(this.name);
        String name;
        while (true) {
            if (!registeredNames.contains(name = generateName())) break;
        }
        registeredNames.add(name);
        this.name = name;
    }
    private static String randomString(char fromChar, char toChar, int len) {
        return random.ints(fromChar, toChar + 1).limit(len)
                .mapToObj(ch -> Character.toString((char)ch))
                .collect(Collectors.joining());
    }
    private static String generateName() {
        return randomString('A', 'Z', 2) + randomString('0', '9', 3);
    }
}